posthog/temporal/workflows/squash_person_overrides.py

```diff
@@ -0,0 +1,476 @@
+    max(version) AS latest_version,
+    min(oldest_event) AS oldest_event_at
+FROM
+    {database}.person_overrides
+WHERE
+    created_at <= %(latest_created_at)s
+GROUP BY
+    team_id, old_person_id
+"""
+
+SELECT_LATEST_CREATED_AT_QUERY = """
+SELECT
+    max(created_at)
+FROM {database}.person_overrides;
+"""
+
+CREATE_JOIN_TABLE_QUERY = """
+CREATE OR REPLACE TABLE {database}.{table_name} (
+    team_id INT NOT NULL,
+    old_person_id UUID NOT NULL,
+    override_person_id UUID NOT NULL
+)
+ENGINE = Join(ANY, LEFT, team_id, old_person_id);
+"""
+
+INSERT_INTO_JOIN_TABLE_QUERY = """
+INSERT INTO {database}.{table_name}
+    SELECT
+        team_id,
+        old_person_id,
+        argMax(override_person_id, merged_at) AS override_person_id
+    FROM {database}.person_overrides
+    WHERE merged_at <= %(latest_created_at)s
+    GROUP BY team_id, old_person_id;
+"""
+
+SQUASH_EVENTS_QUERY = """
+ALTER TABLE
+    {database}.sharded_events
+UPDATE
+    person_id = joinGet({database}.{table_name}, 'override_person_id', toInt32(team_id), person_id)
+IN PARTITION
+    %(partition_id)s
+WHERE
+    true;
+"""
+
+DROP_JOIN_TABLE_QUERY = """
+DROP TABLE {database}.{table_name};
+"""
+
+DELETE_SQUASHED_PERSON_OVERRIDES_QUERY = """
+ALTER TABLE
+    {database}.person_overrides
+DELETE WHERE
+    team_id = %(team_id)s
+    AND old_person_id = %(old_person_id)s
+    AND merged_at <= %(latest_created_at)s;
+"""
+
+SELECT_CREATED_AT_FOR_PERSON_EVENT_QUERY = """
+SELECT
+    min(created_at) AS oldest_event_at
+FROM
+    {database}.sharded_events
+WHERE
+    team_id = %(team_id)s
+    AND person_id = %(old_person_id)s
+    -- Not necessary, but can speed up query.
+    AND created_at <= %(oldest_event_at)s
+    AND created_at >= 0;
+"""
+
+DELETE_FROM_PERSON_OVERRIDE_TABLES = """
+WITH deleted AS (
+    DELETE FROM
+        posthog_personoverridemapping
+    WHERE
+        team_id = %(team_id)s
+        AND uuid = %(old_person_id)s
+        -- We need the id from the mapping table to delete the override.
+        -- So we delete from personoverridemapping before deleting from personoverride.
+        -- But we also need to check personoverride to ensure that the version we are
+        -- deleting is the latest.
+        AND EXISTS (
+            SELECT 1
+            FROM posthog_personoverride
+            WHERE
+                team_id = %(team_id)s
+                AND old_person_id = posthog_personoverridemapping.id
+                AND version = %(latest_version)s
+        )
+    RETURNING id
+)
+DELETE FROM
+    posthog_personoverride
+USING
+    deleted
+WHERE
+    team_id = %(team_id)s
+    AND old_person_id = deleted.id
+    AND version = %(latest_version)s;
+"""
+
+
+class PersonOverrideToDelete(NamedTuple):
+    """A person override that should be deleted after squashing."""
+
+    team_id: int
+    old_person_id: UUID
+    latest_created_at: datetime
+    latest_version: int
+    oldest_event_at: datetime
```

Comment:
just the version should be sufficient here - the others are heuristics and we shouldn't relay on them

Reply:
Exactly, we use them as heuristics. The "absolute oldest event at" is taken from the events table.